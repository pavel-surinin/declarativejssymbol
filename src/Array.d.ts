import { a } from './Symbol'

declare global {
    interface Array<T> {
        [a](): {
            /**
             * docs
             */
            present(): Array<NonNullable<T>>
            /**
             * docs unique by
             */
            uniqueBy<K extends keyof T>(key: K): Array<T>
            uniqueBy<K extends keyof T>(key: StringGetter<T>): Array<T>
            /**
             * docs1
             */
            toObject<K>(key: StringGetter<T>, value: Getter<T, K>): Indexed<T>
            /**
             * docs2
             */
            toObject<K>(key: StringGetter<T>): Indexed<T>
        }
    }

}

