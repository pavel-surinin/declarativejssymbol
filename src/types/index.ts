export * from './ArrayTypes'
export * from './ObjectTypes'
export * from './CommonTypes'