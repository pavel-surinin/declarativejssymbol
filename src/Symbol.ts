export const extend: unique symbol = Symbol('declarativejs.extend')
import './Init'