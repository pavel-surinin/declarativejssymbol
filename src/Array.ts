import { extend } from './Symbol'
import { StringGetter, Getter, Indexed, Predicate } from './Functional'
import { toBe, Reducer, Sort } from 'declarative-js'

export type ArrayExtension<T> = {
    /**
     * Filters out items, that are not equal to provided item in parameters.
     * Objects are compared to be deep equal.
     */
    equal(valueToMatch: T): Array<T>
    /**
     * Filters out items, that are equal to provided item in parameters.
     * Objects are compared to be deep equal.
     */
    notEqual(valueToMatch: T): Array<T>
    /**
     * Determines uniqueness of an objects in array. Be aware that if value
     * is not primitive, deep object equality will be checked to determine
     * uniqueness.
     * @example
     * [1, 2, 2][extend]().unique()
     * // [1, 2]
     * [{a: 1}, {a: 1}, {a: 2}][extend]().unique()
     * // [{a: 1}, {a: 2}]
     */
    unique(): Array<T>
    /** 
     * Filters out items that are empty
     * @example
     * [1, '', [], {}][extend]().notEmpty()
     * // [1]
     */
    notEmpty(): Array<T>
    /** 
     * Filters out items that are not present ({@code undefined} 
     * and {@code null}) in array
     */
    present(): Array<NonNullable<T>>
    /**
     * Determines uniqueness objects key from callback. This value must be
     * comparable with strict equals
     * @param { string } key     callback to resolve comparable value
     * @example
     * [
     *  { title: 'Predator', genre: 'scy-fy' },
     *  { title: 'Predator 2', genre: 'scy-fy' },
     *  { title: 'Alien vs Predator', genre: 'scy-fy' },
     *  { title: 'Tom & Jerry', genre: 'cartoon' },
     * ][extend]()
     *      .uniqueBy('genre')
     * // [
     * //  { title: 'Predator', genre: 'scy-fy' },
     * //  { title: 'Tom & Jerry', genre: 'cartoon' }
     * // ]
     */
    uniqueBy<K extends keyof T>(key: K): Array<T>
    /**
     * Determines uniqueness by value from callback. This value must be
     * comparable with strict equals
     * @param { Function } getValue     callback to resolve comparable value
     * @example
     * [
     *  { title: 'Predator', genre: 'scy-fy' },
     *  { title: 'Predator 2', genre: 'scy-fy' },
     *  { title: 'Alien vs Predator', genre: 'scy-fy' },
     *  { title: 'Tom & Jerry', genre: 'cartoon' },
     * ][extend]()
     *      .uniqueBy(movie => movie.genre)
     * // [
     * //  { title: 'Predator', genre: 'scy-fy' },
     * //  { title: 'Tom & Jerry', genre: 'cartoon' }
     * // ]
     */
    uniqueBy(key: StringGetter<T>): Array<T>
    /**
     * It will pass items from array, while predicate matches. When predicate
     * returns {@code false} none of the items will pass.
     *
     * @param {Function} predicate callback function that returns boolean
     * @example
     * [
     *  { title: 'Predator', genre: 'scy-fy' },
     *  { title: 'Predator 2', genre: 'scy-fy'},
     *  { title: 'Tom & Jerry', genre: 'cartoon' },
     *  { title: 'Alien vs Predator', genre: 'scy-fy' },
     * ][extend]()
     *      .takeWhile(film => film.genre === 'scy-fy')
     * // [
     * //  { title: 'Predator', genre: 'scy-fy' },
     * //  { title: 'Predator 2', genre: 'scy-fy' }
     * // ]
     */
    takeWhile(predicate: Predicate<T>): Array<T>
    /**
     * Function to group by provided key.
     * @param {string}  key     objects key to resolve value,to group by it
     * @throws {Error}          if resolved key from callback is not a string
     * @example
     * [
     *  { title: 'Predator', genre: 'scy-fy },
     *  { title: 'Predator 2', genre: 'scy-fy},
     *  { title: 'Alien vs Predator', genre: 'scy-fy },
     *  { title: 'Tom & Jerry', genre: 'cartoon },
     * ][extend]()
     *      .groupBy('genre')
     */
    groupBy<K extends keyof T>(key: K): Indexed<T[]>
    /**
     * Function to group by provided callback value.
     * @param {Function} getKey              callback to resolve key,to group by it
     * @throws {Error}                       if resolved key from callback is not a string
     * @example
     * [
     *  { title: 'Predator', genre: 'scy-fy },
     *  { title: 'Predator 2', genre: 'scy-fy},
     *  { title: 'Alien vs Predator', genre: 'scy-fy },
     *  { title: 'Tom & Jerry', genre: 'cartoon },
     * ][extend]()
     *      .groupBy(movie => movie.genre)
     */
    groupBy(getKey: StringGetter<T>): Indexed<T[]>
    /**
     * Function to make from 2d array simple array
     * @example
     * [[1,2],[3,4]][extend]().flat() // [1,2,3,4]
     */
    flat(): Array<T>
    /**
     * Collects items to object by key from callback. If function resolves key,
     * that already exists it will throw an Error. Second callback is value mapper.
     * @param {Function} getKey             callback to get key from value
     * @param {Function} getValue           callback to get value to put in object
     * @throws {Error}                      if map has duplicate keys will thrown error
     * @throws {Error}                      if resolved key from callback is not a string      *
     * @example
     * [
     *  { title: 'Predator', genre: 'scy-fy },
     *  { title: 'Predator 2', genre: 'scy-fy},
     *  { title: 'Alien vs Predator', genre: 'scy-fy },
     *  { title: 'Tom & Jerry', genre: 'cartoon },
     * ][extend]()
     *    .toObject(movie => movie.title, movie => movie.genre)
     */
    toObject<K>(getKey: StringGetter<T>, getValue: Getter<T, K>): Indexed<K>
    /**
     * Collects items to object by key from callback. If function resolves
     * key, that already exists it will throw an Error
     * @param {Function} key                  callback to get key from value
     * @throws {Error}                        if map has duplicate keys will thrown error
     * @throws {Error}                        if resolved key from callback is not a string      *   *
     * @example
     * [
     *  { title: 'Predator', genre: 'scy-fy },
     *  { title: 'Predator 2', genre: 'scy-fy},
     *  { title: 'Alien vs Predator', genre: 'scy-fy },
     *  { title: 'Tom & Jerry', genre: 'cartoon },
     * ][extend]()
     *      .toObject(movie => movie.title)
     */
    toObject(key: StringGetter<T>): Indexed<T>
    /**
     * Reduces array of objects to one object, There is three merge strategies
     * @see Reducer.MergeStrategy (declarative-js)
     * @param merge {@link MergeStrategy} = default is OVERRIDE
     */
    merge<R extends object>(strategy?: Reducer.MergeStrategy): T & R
    /**
     * Sorts array in ascending order by values provided from callbacks.
     * First callback has highest priority in sorting and so on.
     * It accepts as many callbacks as You need.
     * @param {...Function} getters     functions to get values to be compared
     * @returns a closure that can be used in array.sort() function
     * @example
     * persons[extend]().ascendingBy(
     *       x => x.name,
     *       x => x.lastName,
     *       x => x.age
     *   )
     * // [
     * //  { name: 'andrew', lastName: 'Aa', age: 1 },
     * //  { name: 'andrew', lastName: 'Bb', age: 1 },
     * //  { name: 'andrew', lastName: 'Bb', age: 2 },
     * //  { name: 'billy', lastName: 'Cc', age: 1 },
     * //  { name: 'billy', lastName: 'Cc', age: 5 },
     * // ]
     */
    ascendingBy(...getters: ((val: T) => string | number)[]): Array<T>
    /**
     * Sorts array in ascending order by values resolved from be key.
     * First resolved value has highest priority in sorting and so on.
     * It accepts as many keys as You need.
     * @param {...string} keys     functions to get values to be compared
     * @returns a closure that can be used in array.sort() function
     * @example
     * persons[extend]().ascendingBy(
     *       'name',
     *       'lastName',
     *       'age'
     *   )
     * // [
     * //  { name: 'andrew', lastName: 'Aa', age: 1 },
     * //  { name: 'andrew', lastName: 'Bb', age: 1 },
     * //  { name: 'andrew', lastName: 'Bb', age: 2 },
     * //  { name: 'billy', lastName: 'Cc', age: 1 },
     * //  { name: 'billy', lastName: 'Cc', age: 5 },
     * // ]
     */
    ascendingBy<K extends keyof T>(...keys: K[]): Array<T>
    /**
     * Sorts array in descending order by values provided from callbacks.
     * First callback has highest priority in sorting and so on.
     * It accepts as many callbacks as You need.
     * @param {...Function} getters     functions to get values to be compared
     * @returns a closure that can be used in array.sort() function
     * @example
     * persons[extend]().ascendingBy(
     *       x => x.name,
     *       x => x.lastName,
     *       x => x.age
     *   )
     * // [
     * //  { name: 'billy', lastName: 'Cc', age: 5 },
     * //  { name: 'billy', lastName: 'Cc', age: 1 },
     * //  { name: 'andrew', lastName: 'Bb', age: 2 },
     * //  { name: 'andrew', lastName: 'Bb', age: 1 },
     * //  { name: 'andrew', lastName: 'Aa', age: 1 },
     * // ]
     */
    descendingBy(...getters: ((val: T) => string | number)[]): Array<T>
    /**
     * Sorts array in descending order by values resolved from be key.
     * First resolved value has highest priority in sorting and so on.
     * It accepts as many keys as You need.
     * @param {...string} keys     functions to get values to be compared
     * @returns a closure that can be used in array.sort() function
     * @example
     * persons[extend]().ascendingBy(
     *       'name',
     *       'lastName',
     *       'age'
     *   )
     * // [
     * //  { name: 'billy', lastName: 'Cc', age: 5 },
     * //  { name: 'billy', lastName: 'Cc', age: 1 },
     * //  { name: 'andrew', lastName: 'Bb', age: 2 },
     * //  { name: 'andrew', lastName: 'Bb', age: 1 },
     * //  { name: 'andrew', lastName: 'Aa', age: 1 },
     * // ]
     */
    descendingBy<K extends keyof T>(...keys: K[]): Array<T>
    /**
     * Function that will sort items in array with custom values, by provided order.
     * It accepts as a parameter object with valueToOrderElement mapper and array of custom order rule
     * @param {toValue: function(T): R, R[]} ...conditions
     * @example
     * const result =
     *       testTodoData[extend]().sortBy(
     *           { toValue: x => x.severity, order: ['low', 'medium', 'high'] },
     *           { toValue: x => x.task, order: ['Sleep', 'Drink'] }
     *       )
     *      // { task: 'Sleep', severity: 'low' },
     *      // { task: 'Drink', severity: 'low' },
     *      // { task: 'Eat', severity: 'medium' },
     *      // { task: 'Code', severity: 'high' },
     */
    sortBy(...conditions: Sort.SortingCondition<T, any>[]): Array<T>
    /**
     * Function that will sort items in array with custom values, by provided order.
     * It accepts as a parameter object with valueToOrderElement mapper and array of custom order rule
     * @param {toValue: function(T): R, R[]} ...conditions
     * @example
     * const result =
     *       testTodoData[extend]().sortBy('severity', ['low', 'medium', 'high'])
     *      // { task: 'Sleep', severity: 'low' },
     *      // { task: 'Eat', severity: 'medium' },
     *      // { task: 'Code', severity: 'high' },
     */
    sortBy<K extends keyof T>(key: K, values: T[K][]): Array<T>
    /**
     * Function that will sort items in array, by provided order.
     * It accepts as a parameter array of custom order rule.
     * Element, that are not present in order array will be at he the end of the sorted list.
     * @param order     array of custom order of items that are being sorted.
     * @example
     * const testData =
     *   ['bar', 'medium', 'foo', 'low']
     * const result =
     *   testData[extend]()
     *      .orderedBy(['low', 'medium', 'high'])
     * // ['low', 'medium', 'bar', 'foo', ]
     */
    orderedBy(order: Array<T>): Array<T>
}

declare global {
    interface Array<T> {
        [extend](): ArrayExtension<T>
    }
}

export const arrayFx: ArrayExtension<any> = {
    present<T>(this: Array<T>): Array<NonNullable<T>> {
        return this.filter(toBe.present) as Array<NonNullable<T>>
    },
    equal<T>(this: Array<T>, valueToMatch: T): Array<T> {
        return this.filter(toBe.equal(valueToMatch))
    },
    notEqual<T>(this: Array<T>, valueToMatch: T): Array<T> {
        return this.filter(toBe.notEqual(valueToMatch))
    },
    notEmpty<T>(this: Array<T>): Array<T> {
        return this.filter(toBe.notEmpty)
    },
    uniqueBy<T>(this: Array<T>, key: any): Array<T> {
        return this.filter(toBe.uniqueBy(key))
    },
    unique<T>(this: Array<T>): Array<T> {
        return this.filter(toBe.unique)
    },
    takeWhile<T>(this: Array<T>, predicate: Predicate<T>): Array<T> {
        return this.filter(toBe.takeWhile(predicate))
    },
    toObject<T, K>(this: Array<T>, key: StringGetter<T>, value?: Getter<T, K>): Indexed<T> {
        return this.reduce(Reducer.toObject(key, value!), {}) as Indexed<T>
    },
    groupBy<T>(this: Array<T>, key: any): Indexed<T[]> {
        return this.reduce(Reducer.groupBy(key), Reducer.Map()).toObject()
    },
    flat<T, A>(this: Array<A extends Array<T> ? Array<T> : never>): Array<T> {
        return this.reduce(Reducer.flat, [])
    },
    merge<T extends object, R extends object>(this: Array<T>, strategy?: Reducer.MergeStrategy): T & R {
        return this.reduce(
            (pv: T & R, cv: T) => Reducer.toMergedObject(strategy)(pv, cv),
            {} as T & R
        )
    },
    ascendingBy<T>(this: Array<T>, ...getters: any[]): Array<T> {
        return this.sort(Sort.ascendingBy(...getters))
    },
    descendingBy<T>(this: Array<T>, ...getters: any[]): Array<T> {
        return this.sort(Sort.descendingBy(...getters))
    },
    orderedBy<T>(this: Array<T>, order: Array<T>): Array<T> {
        return this.sort(Sort.orderedBy(order))
    },
    sortBy<T>(this: Array<T>, ...conditions: any[]): Array<T> {
        return this.sort(Sort.by(...conditions))
    }
}