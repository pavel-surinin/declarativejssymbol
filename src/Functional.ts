import { ObjectExtension } from './Object'
import { extend } from './Symbol'

export type StringGetter<T> = Getter<T, string>
export type Getter<O, V> = (value: O) => V
export type Predicate<T> = (vlaue: T) => boolean
export interface Indexed<V> {
    [key: string]: V
    [extend](): ObjectExtension<V>
}
