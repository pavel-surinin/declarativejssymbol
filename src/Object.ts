import { extend } from './Symbol'
import { Indexed } from './Functional'

declare global {
    interface Object {
        [extend](): ObjectExtension<any>
    }
}

export type ObjectExtension<T> = {
    /** Returns object keys */
    keys(): string[]
    /** Returns object values */
    values(): T[]
    /** Returns object entries as a {key, value} objects array */
    entries(): ObjectEntry<T>[]
    /** 
     * Is object contains key
     * @param {string} key to search
     * @returns {boolean} 
     */
    containsKey(key: string): boolean
    /** 
     * Is object contains value
     * @param {T} value to search
     * @returns {boolean} 
     */
    containsValue(value: T): boolean
}

export interface ObjectEntry<T> {
    key: PropertyKey,
    value: T
}

export const objectFx: ObjectExtension<any> = {
    containsKey(this: Indexed<any>, key: PropertyKey): boolean {
        return Object.keys(this).some(k => k === key)
    },
    containsValue(this: Indexed<any>, value: any): boolean {
        return Object.keys(this).some(k => this[k] === value)
    },
    keys(this: Indexed<any>): string[] {
        return Object.keys(this)
    },
    values<T>(this: Indexed<any>): T[] {
        return Object.keys(this).map(k => (this as any)[k])
    },
    entries<T>(this: Indexed<any>): ObjectEntry<T>[] {
        return Object.keys(this).map(k => ({ key: k, value: this[k] }))
    }
}
