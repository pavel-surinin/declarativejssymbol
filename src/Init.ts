import { extend } from './Symbol'
import { Reducer } from 'declarative-js'
import { Indexed } from './Functional'
import { ObjectExtension, objectFx } from './Object'
import { ArrayExtension, arrayFx } from './Array'

function bindFunctions(functions: Indexed<Function>, scope: any) {
    const ex = Object.keys(functions).map(k => functions[k])
        .reduce(
            Reducer.toObject(
                fx => (fx as Function).name,
                fx => (fx as Function).bind(scope)),
            {}
        )
    return ex
}

(function () {
    const keys = Reflect.ownKeys(Object.prototype)
    const isExtended = keys.some(k => k === extend)
    if (!isExtended) {
        Array.prototype[extend] = function extendArrayPrototypes() {
            return bindFunctions(arrayFx, this) as ArrayExtension<any>
        }
        Object.prototype[extend] = function extendObjectPrototypes() {
            return bindFunctions(objectFx, this) as ObjectExtension<any>
        }
    }
})()
