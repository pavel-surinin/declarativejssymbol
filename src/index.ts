import './Init'
import './Array'
import './Functional'
import './Symbol'