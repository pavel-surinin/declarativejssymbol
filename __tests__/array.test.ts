import { extend } from '../src/Symbol'

interface A {
    person?: P
}
interface P {
    name: string
}
const arr: A[] = [
    { person: { name: 'John' } },
    { person: { name: 'John' } },
    { person: { name: 'Mike' } },
    { person: undefined },
    { person: { name: 'Mike' } },
    { person: { name: 'Mike' } },
    { person: undefined },
    { person: { name: 'Peter' } },
    { person: { name: 'Peter' } },
    { person: { name: 'Peter' } }
]
describe('extend', () => {
    describe('array', () => {
        describe('sort', () => {
            const tasks = [
                { task: 'Sleep', severity: 'low' },
                { task: 'Eat', severity: 'medium' },
                { task: 'Drink', severity: 'low' },
                { task: 'Code', severity: 'high' }
            ]
            it('sortBy key', () => {
                const sorted = tasks[extend]()
                    .sortBy('severity', ['high', 'low', 'medium'])
                expect(sorted).toMatchObject([
                    { task: 'Code', severity: 'high' },
                    { task: 'Sleep', severity: 'low' },
                    { task: 'Drink', severity: 'low' },
                    { task: 'Eat', severity: 'medium' },
                ])
            })
            it('sortBy callback', () => {
                const sorted = tasks[extend]()
                    .sortBy(
                        {
                            toValue: x => x.severity,
                            order: ['high', 'low', 'medium']
                        },
                        {
                            toValue: x => x.task,
                            order: ['Drink', 'Sleep']
                        }
                    )
                expect(sorted).toMatchObject([
                    { task: 'Code', severity: 'high' },
                    { task: 'Drink', severity: 'low' },
                    { task: 'Sleep', severity: 'low' },
                    { task: 'Eat', severity: 'medium' },
                ])
            })
            it('orderedBy', () => {
                const sorted = tasks
                    .map(x => x.severity)[extend]()
                    .orderedBy(['high', 'medium', 'low'])
                expect(sorted).toMatchObject(['high', 'medium', 'low', 'low'])
            })
            it('ascending callback', () => {
                const sorted = tasks[extend]()
                    .ascendingBy(x => x.severity, x => x.task)
                expect(sorted).toMatchObject([
                    { task: 'Code', severity: 'high' },
                    { task: 'Drink', severity: 'low' },
                    { task: 'Sleep', severity: 'low' },
                    { task: 'Eat', severity: 'medium' },
                ])
            })
            it('ascending key', () => {
                const sorted = tasks[extend]()
                    .ascendingBy('severity', 'task')
                expect(sorted).toMatchObject([
                    { task: 'Code', severity: 'high' },
                    { task: 'Drink', severity: 'low' },
                    { task: 'Sleep', severity: 'low' },
                    { task: 'Eat', severity: 'medium' },
                ])
            })
        })
        it('not equal', () => {
            const peter = { name: 'Peter' }
            const result = arr
                .map(x => x.person)[extend]()
                .notEqual(peter)
            expect(result).toHaveLength(7)
        })
        it('unique', () => {
            const result = arr[extend]().unique()
            expect(result).toHaveLength(4)
        })
        it('unique by key', () => {
            const result = arr
                .map(x => x.person)[extend]()
                .present()[extend]()
                .uniqueBy('name')
            expect(result).toHaveLength(3)
        })
        it('unique by cb', () => {
            const result = arr
                .map(x => x.person)[extend]()
                .present()[extend]()
                .uniqueBy(x => x.name)
            expect(result).toHaveLength(3)
        })
        it('group by cb', () => {
            const result = arr
                .map(x => x.person)[extend]()
                .present()[extend]()
                .groupBy(x => x.name)
            expect(result).toMatchSnapshot()
        })
        it('group by key', () => {
            const result = arr
                .map(x => x.person)[extend]()
                .present()[extend]()
                .groupBy('name')
            expect(result).toMatchSnapshot()
        })
        it('notEmpty', () => {
            expect([1, '', [], {}][extend]().notEmpty()).toHaveLength(1)
        })
        it('takeWhile', () => {
            expect([1, 2, 3, 4][extend]().takeWhile(n => n !== 3)).toHaveLength(2)
        })
        it('flat', () => {
            expect([[1], [2]][extend]().flat()).toMatchObject([1, 2])
        })
        it('merge', () => {
            const array = [
                { a: 1 },
                { b: 2 }
            ]
            expect(array[extend]().merge()).toMatchObject({ a: 1, b: 2 })
        })
    })
    it('general', () => {
        const result = arr
            .map(x => x.person)[extend]()
            .present()[extend]()
            .uniqueBy(x => x.name)
            .filter(x => x.name !== 'John')[extend]()
            .toObject(x => x.name, x => x.name)[extend]()
            .values()
            .sort()
        expect(result).toMatchObject(['Peter', 'Mike'].sort())
    })
})
