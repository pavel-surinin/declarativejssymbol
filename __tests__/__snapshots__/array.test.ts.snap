// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extend array group by cb 1`] = `
Object {
  "John": Array [
    Object {
      "name": "John",
    },
    Object {
      "name": "John",
    },
  ],
  "Mike": Array [
    Object {
      "name": "Mike",
    },
    Object {
      "name": "Mike",
    },
    Object {
      "name": "Mike",
    },
  ],
  "Peter": Array [
    Object {
      "name": "Peter",
    },
    Object {
      "name": "Peter",
    },
    Object {
      "name": "Peter",
    },
  ],
}
`;

exports[`extend array group by key 1`] = `
Object {
  "John": Array [
    Object {
      "name": "John",
    },
    Object {
      "name": "John",
    },
  ],
  "Mike": Array [
    Object {
      "name": "Mike",
    },
    Object {
      "name": "Mike",
    },
    Object {
      "name": "Mike",
    },
  ],
  "Peter": Array [
    Object {
      "name": "Peter",
    },
    Object {
      "name": "Peter",
    },
    Object {
      "name": "Peter",
    },
  ],
}
`;
